kernel/pipes.c

```diff
@@ -420,7 +418,9 @@ int z_impl_k_pipe_put(struct k_pipe *pipe, void *data, size_t bytes_to_write,
 	*bytes_written = pipe_write(pipe, &src_list,
 				    &dest_list, &reschedule_needed);
 
-	if (*bytes_written) {
+	/* Only handle poll events if the pipe has an underlying buffer with some bytes in it */
+	
+	if (pipe->buffer != NULL && pipe->bytes_used) {
```

Comment:
I think this is better, but still not quite right. I am basing on this assessment on a comparison to how message queues handle events. Granted, the nature of pipes is a little different than message queues but I think that we can draw some parallels in expected behaviour.

I observe that in k_msgq_put() the handling of poll events is only done if a message was added to its message buffer. If the message buffer is empty and/or the message had to pend, then we handle a poll event. If I follow the proposed logic here in k_pipe_put() correctly, then we are handling a poll event even if the buffer was full and nothing was written into it.

What do you think of the following ...

```
    if ((pipe->bytes_used != 0) && (*bytes_written != 0)) {
        handle_poll_events(pipe);
    }
```
I might have missed/forgotten something, but I think that *bytes_used* if non-zero implies that *buffer* is also non-NULL. And if both conditions are true, then something was definitely written into the pipe buffer, which in turn would require the handling of a poll event.

Reply:
I forgot to use 0U instead of 0 in the example above. :)

Reply:
To be honest with you, I have played around with many forms of these various conditions already ;) 

I think that really it just comes down to having behaviour that is reasonable and expected (and well documented), and I think you make a good argument about what the condition should be as msgq already does this (I didn't think to look there). 

I am also trying to be aware of the fact that I am fairly new to the inner depths of Zephyr, and so I was hoping that someone (like yourself) could guide me on what you expected was reasonable.